# Exercise 02 -  OLAP Cubes - Grouping Sets

All the databases table in this demo are based on public database samples and transformations
- `Sakila` is a sample database created by `MySql` [Link](https://dev.mysql.com/doc/sakila/en/sakila-structure.html)
- The postgresql version of it is called `Pagila` [Link](https://github.com/devrimgunduz/pagila)
- The facts and dimension tables design is based on O'Reilly's public dimensional modelling tutorial schema [Link](http://archive.oreilly.com/oreillyschool/courses/dba3/index.html)

Start by connecting to the database by running the cells below. If you are coming back to this exercise, then uncomment and run the first cell to recreate the database. If you recently completed the slicing and dicing exercise, then skip to the second cell.

In [ ]:
# !PGPASSWORD=student createdb -h 127.0.0.1 -U student pagila_star
# !PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila_star -f Data/pagila-star.sql

### Connect to the local database where Pagila is loaded

In [1]:
import sql
%load_ext sql

DB_ENDPOINT = "127.0.0.1"
DB = 'pagila_star'
DB_USER = 'student'
DB_PASSWORD = 'student'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)
%sql $conn_string

postgresql://student:student@127.0.0.1:5432/pagila_star


'Connected: student@pagila_star'

### Star Schema

<img src="pagila-star.png" width="50%"/>

# Grouping Sets
- It happens often that for 3 dimensions, you want to aggregate a fact:
    - by nothing (total)
    - then by the 1st dimension
    - then by the 2nd 
    - then by the 3rd 
    - then by the 1st and 2nd
    - then by the 2nd and 3rd
    - then by the 1st and 3rd
    - then by the 1st and 2nd and 3rd
    
- Since this is very common, and in all cases, we are iterating through all the fact table anyhow, there is a more clever way to do that using the SQL grouping statement "GROUPING SETS" 

## Total Revenue

TODO: Write a query that calculates total revenue (sales_amount)

In [2]:
%%sql
SELECT sum(sales_amount)
FROM factSales

 * postgresql://student:***@127.0.0.1:5432/pagila_star
1 rows affected.


[(Decimal('67416.51'),)]

## Revenue by Country
TODO: Write a query that calculates total revenue (sales_amount) by country

In [3]:
%%sql
SELECT sum(sales_amount), dc.country
FROM factSales fs
    JOIN dimCustomer dc ON(fs.customer_key = dc.customer_key)
GROUP BY dc.country

 * postgresql://student:***@127.0.0.1:5432/pagila_star
108 rows affected.


[(Decimal('191.47'), 'Cambodia'),
 (Decimal('1662.12'), 'Turkey'),
 (Decimal('831.04'), 'Germany'),
 (Decimal('135.68'), 'Chad'),
 (Decimal('93.78'), 'Madagascar'),
 (Decimal('374.04'), 'France'),
 (Decimal('111.73'), 'Iraq'),
 (Decimal('709.41'), 'Colombia'),
 (Decimal('3471.74'), 'Japan'),
 (Decimal('71.80'), 'American Samoa'),
 (Decimal('122.68'), 'Virgin Islands, U.S.'),
 (Decimal('746.28'), 'Vietnam'),
 (Decimal('111.74'), 'Kuwait'),
 (Decimal('390.13'), 'Ecuador'),
 (Decimal('116.78'), 'Nepal'),
 (Decimal('200.46'), 'Cameroon'),
 (Decimal('523.79'), 'Saudi Arabia'),
 (Decimal('402.05'), 'Bangladesh'),
 (Decimal('73.76'), 'Lithuania'),
 (Decimal('586.66'), 'Netherlands'),
 (Decimal('259.43'), 'Yugoslavia'),
 (Decimal('129.70'), 'Gambia'),
 (Decimal('1511.48'), 'Nigeria'),
 (Decimal('318.23'), 'Dominican Republic'),
 (Decimal('114.72'), 'Faroe Islands'),
 (Decimal('694.39'), 'Egypt'),
 (Decimal('134.67'), 'Zambia'),
 (Decimal('4110.32'), 'United States'),
 (Decimal('1434.48'), 'Argentina'),
 (Decimal('89.74'), 'Slovakia'),
 (Decimal('215.48'), 'Angola'),
 (Decimal('3045.87'), 'Russian Federation'),
 (Decimal('187.50'), 'Oman'),
 (Decimal('136.73'), 'Turkmenistan'),
 (Decimal('67.82'), 'Afghanistan'),
 (Decimal('133.71'), 'Czech Republic'),
 (Decimal('114.72'), 'Liechtenstein'),
 (Decimal('91.77'), 'Ethiopia'),
 (Decimal('245.43'), 'Azerbaijan'),
 (Decimal('3200.52'), 'Brazil'),
 (Decimal('422.01'), 'Israel'),
 (Decimal('100.75'), 'Senegal'),
 (Decimal('328.29'), 'Chile'),
 (Decimal('92.76'), 'New Zealand'),
 (Decimal('111.71'), 'Hungary'),
 (Decimal('344.20'), 'Mozambique'),
 (Decimal('3307.04'), 'Mexico'),
 (Decimal('101.74'), 'Finland'),
 (Decimal('1210.94'), 'Taiwan'),
 (Decimal('116.70'), 'Sri Lanka'),
 (Decimal('419.04'), 'Thailand'),
 (Decimal('115.70'), 'Estonia'),
 (Decimal('262.40'), 'Latvia'),
 (Decimal('683.30'), 'Venezuela'),
 (Decimal('468.88'), 'Peru'),
 (Decimal('152.66'), 'Holy See (Vatican City State)'),
 (Decimal('216.54'), 'Runion'),
 (Decimal('106.65'), 'Anguilla'),
 (Decimal('112.75'), 'Bahrain'),
 (Decimal('383.10'), 'Algeria'),
 (Decimal('369.15'), 'Malaysia'),
 (Decimal('950.75'), 'Iran'),
 (Decimal('606.58'), 'Spain'),
 (Decimal('730.42'), 'Ukraine'),
 (Decimal('73.82'), 'Tonga'),
 (Decimal('574.65'), 'South Korea'),
 (Decimal('277.34'), 'Belarus'),
 (Decimal('103.78'), 'French Guiana'),
 (Decimal('142.70'), 'Hong Kong'),
 (Decimal('123.72'), 'Malawi'),
 (Decimal('148.69'), 'Nauru'),
 (Decimal('235.38'), 'Romania'),
 (Decimal('113.69'), 'North Korea'),
 (Decimal('254.39'), 'Puerto Rico'),
 (Decimal('593.63'), 'Canada'),
 (Decimal('5802.73'), 'China'),
 (Decimal('307.22'), 'Austria'),
 (Decimal('127.66'), 'Moldova'),
 (Decimal('1204.15'), 'South Africa'),
 (Decimal('253.46'), 'Kenya'),
 (Decimal('877.97'), 'Poland'),
 (Decimal('198.48'), 'Kazakstan'),
 (Decimal('924.80'), 'United Kingdom'),
 (Decimal('113.65'), 'Brunei'),
 (Decimal('78.77'), 'Tunisia'),
 (Decimal('831.11'), 'Italy'),
 (Decimal('525.72'), 'Pakistan'),
 (Decimal('144.66'), 'Sweden'),
 (Decimal('118.75'), 'Armenia'),
 (Decimal('232.46'), 'Greece'),
 (Decimal('6630.27'), 'India'),
 (Decimal('510.83'), 'Yemen'),
 (Decimal('275.38'), 'Paraguay'),
 (Decimal('194.48'), 'Myanmar'),
 (Decimal('2381.32'), 'Philippines'),
 (Decimal('307.29'), 'Morocco'),
 (Decimal('137.66'), 'Greenland'),
 (Decimal('252.39'), 'Switzerland'),
 (Decimal('1510.33'), 'Indonesia'),
 (Decimal('120.74'), 'Tuvalu'),
 (Decimal('96.75'), 'Saint Vincent and the Grenadines'),
 (Decimal('183.53'), 'Bolivia'),
 (Decimal('212.50'), 'Congo, The Democratic Republic of the'),
 (Decimal('227.46'), 'Sudan'),
 (Decimal('333.16'), 'United Arab Emirates'),
 (Decimal('341.17'), 'Tanzania'),
 (Decimal('235.46'), 'French Polynesia'),
 (Decimal('204.50'), 'Bulgaria')]

## Revenue by Month
TODO: Write a query that calculates total revenue (sales_amount) by month

In [5]:
%%sql
SELECT sum(sales_amount), dc.month
FROM factSales fs
    JOIN dimDate dc ON(fs.data_key = dc.date_key)
GROUP BY dc.month

 * postgresql://student:***@127.0.0.1:5432/pagila_star
5 rows affected.


[(Decimal('57118.92'), 4),
 (Decimal('1028.36'), 5),
 (Decimal('9648.86'), 1),
 (Decimal('19263.76'), 2),
 (Decimal('47773.12'), 3)]

## Revenue by Month & Country
TODO: Write a query that calculates total revenue (sales_amount) by month and country. Sort the data by month, country, and revenue in descending order. The first few rows of your output should match the table below.

In [11]:
%%sql
SELECT sum(sales_amount) as revenue, dd.month, dc.country
FROM factSales fs
    JOIN dimDate dd ON(fs.data_key = dd.date_key)
    JOIN dimCustomer dc ON(fs.customer_key = dc.customer_key)
GROUP BY dd.month, dc.country
ORDER BY dd.month, dc.country, revenue desc

 * postgresql://student:***@127.0.0.1:5432/pagila_star
484 rows affected.


[(Decimal('67.84'), 1, 'Algeria'),
 (Decimal('55.86'), 1, 'Angola'),
 (Decimal('13.94'), 1, 'Anguilla'),
 (Decimal('163.56'), 1, 'Argentina'),
 (Decimal('45.84'), 1, 'Austria'),
 (Decimal('41.94'), 1, 'Azerbaijan'),
 (Decimal('7.98'), 1, 'Bahrain'),
 (Decimal('39.90'), 1, 'Bangladesh'),
 (Decimal('11.96'), 1, 'Belarus'),
 (Decimal('9.94'), 1, 'Bolivia'),
 (Decimal('421.00'), 1, 'Brazil'),
 (Decimal('11.98'), 1, 'Brunei'),
 (Decimal('19.96'), 1, 'Bulgaria'),
 (Decimal('23.92'), 1, 'Cambodia'),
 (Decimal('27.94'), 1, 'Cameroon'),
 (Decimal('67.86'), 1, 'Canada'),
 (Decimal('25.92'), 1, 'Chad'),
 (Decimal('49.90'), 1, 'Chile'),
 (Decimal('911.82'), 1, 'China'),
 (Decimal('95.74'), 1, 'Colombia'),
 (Decimal('87.84'), 1, 'Congo, The Democratic Republic of the'),
 (Decimal('1.98'), 1, 'Czech Republic'),
 (Decimal('27.94'), 1, 'Dominican Republic'),
 (Decimal('41.90'), 1, 'Ecuador'),
 (Decimal('69.82'), 1, 'Egypt'),
 (Decimal('19.96'), 1, 'Estonia'),
 (Decimal('13.98'), 1, 'Faroe Islands'),
 (Decimal('7.98'), 1, 'Finland'),
 (Decimal('79.84'), 1, 'France'),
 (Decimal('11.96'), 1, 'French Guiana'),
 (Decimal('59.88'), 1, 'French Polynesia'),
 (Decimal('29.94'), 1, 'Gambia'),
 (Decimal('179.60'), 1, 'Germany'),
 (Decimal('55.84'), 1, 'Greece'),
 (Decimal('35.88'), 1, 'Greenland'),
 (Decimal('11.96'), 1, 'Holy See (Vatican City State)'),
 (Decimal('859.78'), 1, 'India'),
 (Decimal('271.38'), 1, 'Indonesia'),
 (Decimal('111.68'), 1, 'Iran'),
 (Decimal('79.78'), 1, 'Israel'),
 (Decimal('103.78'), 1, 'Italy'),
 (Decimal('576.70'), 1, 'Japan'),
 (Decimal('11.94'), 1, 'Kazakstan'),
 (Decimal('15.94'), 1, 'Kenya'),
 (Decimal('9.98'), 1, 'Kuwait'),
 (Decimal('25.94'), 1, 'Latvia'),
 (Decimal('29.96'), 1, 'Liechtenstein'),
 (Decimal('19.96'), 1, 'Lithuania'),
 (Decimal('1.98'), 1, 'Madagascar'),
 (Decimal('3.98'), 1, 'Malawi'),
 (Decimal('61.88'), 1, 'Malaysia'),
 (Decimal('526.74'), 1, 'Mexico'),
 (Decimal('65.88'), 1, 'Morocco'),
 (Decimal('57.90'), 1, 'Mozambique'),
 (Decimal('29.90'), 1, 'Myanmar'),
 (Decimal('9.98'), 1, 'Nauru'),
 (Decimal('31.94'), 1, 'Nepal'),
 (Decimal('51.88'), 1, 'Netherlands'),
 (Decimal('13.98'), 1, 'New Zealand'),
 (Decimal('269.42'), 1, 'Nigeria'),
 (Decimal('11.96'), 1, 'North Korea'),
 (Decimal('51.88'), 1, 'Oman'),
 (Decimal('103.76'), 1, 'Pakistan'),
 (Decimal('3.96'), 1, 'Paraguay'),
 (Decimal('81.82'), 1, 'Peru'),
 (Decimal('323.24'), 1, 'Philippines'),
 (Decimal('109.78'), 1, 'Poland'),
 (Decimal('31.90'), 1, 'Puerto Rico'),
 (Decimal('33.92'), 1, 'Romania'),
 (Decimal('9.98'), 1, 'Runion'),
 (Decimal('450.80'), 1, 'Russian Federation'),
 (Decimal('31.94'), 1, 'Saint Vincent and the Grenadines'),
 (Decimal('131.74'), 1, 'Saudi Arabia'),
 (Decimal('9.98'), 1, 'Senegal'),
 (Decimal('17.94'), 1, 'Slovakia'),
 (Decimal('199.52'), 1, 'South Africa'),
 (Decimal('93.76'), 1, 'South Korea'),
 (Decimal('113.70'), 1, 'Spain'),
 (Decimal('25.94'), 1, 'Sri Lanka'),
 (Decimal('49.90'), 1, 'Sudan'),
 (Decimal('9.98'), 1, 'Sweden'),
 (Decimal('7.96'), 1, 'Switzerland'),
 (Decimal('111.68'), 1, 'Taiwan'),
 (Decimal('37.90'), 1, 'Tanzania'),
 (Decimal('35.92'), 1, 'Thailand'),
 (Decimal('17.96'), 1, 'Tonga'),
 (Decimal('241.46'), 1, 'Turkey'),
 (Decimal('19.98'), 1, 'Turkmenistan'),
 (Decimal('53.92'), 1, 'Tuvalu'),
 (Decimal('109.78'), 1, 'Ukraine'),
 (Decimal('25.90'), 1, 'United Arab Emirates'),
 (Decimal('147.68'), 1, 'United Kingdom'),
 (Decimal('618.60'), 1, 'United States'),
 (Decimal('49.84'), 1, 'Venezuela'),
 (Decimal('67.82'), 1, 'Vietnam'),
 (Decimal('1.98'), 1, 'Virgin Islands, U.S.'),
 (Decimal('31.88'), 1, 'Yemen'),
 (Decimal('51.88'), 1, 'Yugoslavia'),
 (Decimal('25.94'), 1, 'Zambia'),
 (Decimal('9.94'), 2, 'Afghanistan'),
 (Decimal('121.76'), 2, 'Algeria'),
 (Decimal('47.90'), 2, 'American Samoa'),
 (Decimal('71.84'), 2, 'Angola'),
 (Decimal('33.90'), 2, 'Anguilla'),
 (Decimal('438.98'), 2, 'Argentina'),
 (Decimal('25.96'), 2, 'Armenia'),
 (Decimal('79.78'), 2, 'Austria'),
 (Decimal('51.86'), 2, 'Azerbaijan

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>month</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>1</td>
        <td>Australia</td>
        <td>2364.19</td>
    </tr>
    <tr>
        <td>1</td>
        <td>Canada</td>
        <td>2460.24</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Australia</td>
        <td>4895.10</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Canada</td>
        <td>4736.78</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Australia</td>
        <td>12060.33</td>
    </tr>
</tbody></table></div>

## Revenue Total, by Month, by Country, by Month & Country All in one shot

TODO: Write a query that calculates total revenue at the various grouping levels done above (total, by month, by country, by month & country) all at once using the grouping sets function. Your output should match the table below.

In [8]:
%%sql
SELECT sum(sales_amount) as revenue, dd.month, dc.country
FROM factSales fs
    JOIN dimDate dd ON(fs.data_key = dd.date_key)
    JOIN dimCustomer dc ON(fs.customer_key = dc.customer_key)
GROUP BY grouping sets ((), dd.month, dc.country, (dd.month, dc.country))

 * postgresql://student:***@127.0.0.1:5432/pagila_star
598 rows affected.


[(Decimal('67.84'), 1, 'Algeria'),
 (Decimal('55.86'), 1, 'Angola'),
 (Decimal('13.94'), 1, 'Anguilla'),
 (Decimal('163.56'), 1, 'Argentina'),
 (Decimal('45.84'), 1, 'Austria'),
 (Decimal('41.94'), 1, 'Azerbaijan'),
 (Decimal('7.98'), 1, 'Bahrain'),
 (Decimal('39.90'), 1, 'Bangladesh'),
 (Decimal('11.96'), 1, 'Belarus'),
 (Decimal('9.94'), 1, 'Bolivia'),
 (Decimal('421.00'), 1, 'Brazil'),
 (Decimal('11.98'), 1, 'Brunei'),
 (Decimal('19.96'), 1, 'Bulgaria'),
 (Decimal('23.92'), 1, 'Cambodia'),
 (Decimal('27.94'), 1, 'Cameroon'),
 (Decimal('67.86'), 1, 'Canada'),
 (Decimal('25.92'), 1, 'Chad'),
 (Decimal('49.90'), 1, 'Chile'),
 (Decimal('911.82'), 1, 'China'),
 (Decimal('95.74'), 1, 'Colombia'),
 (Decimal('87.84'), 1, 'Congo, The Democratic Republic of the'),
 (Decimal('1.98'), 1, 'Czech Republic'),
 (Decimal('27.94'), 1, 'Dominican Republic'),
 (Decimal('41.90'), 1, 'Ecuador'),
 (Decimal('69.82'), 1, 'Egypt'),
 (Decimal('19.96'), 1, 'Estonia'),
 (Decimal('13.98'), 1, 'Faroe Islands'),
 (Decimal('7.98'), 1, 'Finland'),
 (Decimal('79.84'), 1, 'France'),
 (Decimal('11.96'), 1, 'French Guiana'),
 (Decimal('59.88'), 1, 'French Polynesia'),
 (Decimal('29.94'), 1, 'Gambia'),
 (Decimal('179.60'), 1, 'Germany'),
 (Decimal('55.84'), 1, 'Greece'),
 (Decimal('35.88'), 1, 'Greenland'),
 (Decimal('11.96'), 1, 'Holy See (Vatican City State)'),
 (Decimal('859.78'), 1, 'India'),
 (Decimal('271.38'), 1, 'Indonesia'),
 (Decimal('111.68'), 1, 'Iran'),
 (Decimal('79.78'), 1, 'Israel'),
 (Decimal('103.78'), 1, 'Italy'),
 (Decimal('576.70'), 1, 'Japan'),
 (Decimal('11.94'), 1, 'Kazakstan'),
 (Decimal('15.94'), 1, 'Kenya'),
 (Decimal('9.98'), 1, 'Kuwait'),
 (Decimal('25.94'), 1, 'Latvia'),
 (Decimal('29.96'), 1, 'Liechtenstein'),
 (Decimal('19.96'), 1, 'Lithuania'),
 (Decimal('1.98'), 1, 'Madagascar'),
 (Decimal('3.98'), 1, 'Malawi'),
 (Decimal('61.88'), 1, 'Malaysia'),
 (Decimal('526.74'), 1, 'Mexico'),
 (Decimal('65.88'), 1, 'Morocco'),
 (Decimal('57.90'), 1, 'Mozambique'),
 (Decimal('29.90'), 1, 'Myanmar'),
 (Decimal('9.98'), 1, 'Nauru'),
 (Decimal('31.94'), 1, 'Nepal'),
 (Decimal('51.88'), 1, 'Netherlands'),
 (Decimal('13.98'), 1, 'New Zealand'),
 (Decimal('269.42'), 1, 'Nigeria'),
 (Decimal('11.96'), 1, 'North Korea'),
 (Decimal('51.88'), 1, 'Oman'),
 (Decimal('103.76'), 1, 'Pakistan'),
 (Decimal('3.96'), 1, 'Paraguay'),
 (Decimal('81.82'), 1, 'Peru'),
 (Decimal('323.24'), 1, 'Philippines'),
 (Decimal('109.78'), 1, 'Poland'),
 (Decimal('31.90'), 1, 'Puerto Rico'),
 (Decimal('33.92'), 1, 'Romania'),
 (Decimal('9.98'), 1, 'Runion'),
 (Decimal('450.80'), 1, 'Russian Federation'),
 (Decimal('31.94'), 1, 'Saint Vincent and the Grenadines'),
 (Decimal('131.74'), 1, 'Saudi Arabia'),
 (Decimal('9.98'), 1, 'Senegal'),
 (Decimal('17.94'), 1, 'Slovakia'),
 (Decimal('199.52'), 1, 'South Africa'),
 (Decimal('93.76'), 1, 'South Korea'),
 (Decimal('113.70'), 1, 'Spain'),
 (Decimal('25.94'), 1, 'Sri Lanka'),
 (Decimal('49.90'), 1, 'Sudan'),
 (Decimal('9.98'), 1, 'Sweden'),
 (Decimal('7.96'), 1, 'Switzerland'),
 (Decimal('111.68'), 1, 'Taiwan'),
 (Decimal('37.90'), 1, 'Tanzania'),
 (Decimal('35.92'), 1, 'Thailand'),
 (Decimal('17.96'), 1, 'Tonga'),
 (Decimal('241.46'), 1, 'Turkey'),
 (Decimal('19.98'), 1, 'Turkmenistan'),
 (Decimal('53.92'), 1, 'Tuvalu'),
 (Decimal('109.78'), 1, 'Ukraine'),
 (Decimal('25.90'), 1, 'United Arab Emirates'),
 (Decimal('147.68'), 1, 'United Kingdom'),
 (Decimal('618.60'), 1, 'United States'),
 (Decimal('49.84'), 1, 'Venezuela'),
 (Decimal('67.82'), 1, 'Vietnam'),
 (Decimal('1.98'), 1, 'Virgin Islands, U.S.'),
 (Decimal('31.88'), 1, 'Yemen'),
 (Decimal('51.88'), 1, 'Yugoslavia'),
 (Decimal('25.94'), 1, 'Zambia'),
 (Decimal('9648.86'), 1, None),
 (Decimal('9.94'), 2, 'Afghanistan'),
 (Decimal('121.76'), 2, 'Algeria'),
 (Decimal('47.90'), 2, 'American Samoa'),
 (Decimal('71.84'), 2, 'Angola'),
 (Decimal('33.90'), 2, 'Anguilla'),
 (Decimal('438.98'), 2, 'Argentina'),
 (Decimal('25.96'), 2, 'Armenia'),
 (Decimal('79.78'), 2, 'Austria'),
 (

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>month</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>1</td>
        <td>Australia</td>
        <td>2364.19</td>
    </tr>
    <tr>
        <td>1</td>
        <td>Canada</td>
        <td>2460.24</td>
    </tr>
    <tr>
        <td>1</td>
        <td>None</td>
        <td>4824.43</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Australia</td>
        <td>4895.10</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Canada</td>
        <td>4736.78</td>
    </tr>
    <tr>
        <td>2</td>
        <td>None</td>
        <td>9631.88</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Australia</td>
        <td>12060.33</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Canada</td>
        <td>11826.23</td>
    </tr>
    <tr>
        <td>3</td>
        <td>None</td>
        <td>23886.56</td>
    </tr>
    <tr>
        <td>4</td>
        <td>Australia</td>
        <td>14136.07</td>
    </tr>
    <tr>
        <td>4</td>
        <td>Canada</td>
        <td>14423.39</td>
    </tr>
    <tr>
        <td>4</td>
        <td>None</td>
        <td>28559.46</td>
    </tr>
    <tr>
        <td>5</td>
        <td>Australia</td>
        <td>271.08</td>
    </tr>
    <tr>
        <td>5</td>
        <td>Canada</td>
        <td>243.10</td>
    </tr>
    <tr>
        <td>5</td>
        <td>None</td>
        <td>514.18</td>
    </tr>
    <tr>
        <td>None</td>
        <td>None</td>
        <td>67416.51</td>
    </tr>
    <tr>
        <td>None</td>
        <td>Australia</td>
        <td>33726.77</td>
    </tr>
    <tr>
        <td>None</td>
        <td>Canada</td>
        <td>33689.74</td>
    </tr>
</tbody></table></div>